### PR TITLE
Android permissions JSON export

### PR DESCRIPTION
**Pull request by tuxology**
* We only add permissions that have sources attached. This means there can be `INTERNET` permission assigned, but this is not interesting since its not something where data sources can be linked. `RECORD_AUDIO` however is a valid permissions since it can act as data source.
*  If there are more than once `AndroidManifest.xml` but have similar permissions, we group all permissions and collect their occurrences and group them together as shown below 👇 

Sample JSON:

```json
 "androidPermissions" : [
    {
      "permissionType" : "android.permission.WRITE_EXTERNAL_STORAGE",
      "isUsed" : true,
      "permissionDetail" : {
        "sourceId" : "Data.Sensitive.UserContentData.DeviceStorage",
        "occurrences" : [
          {
            "sample" : "android.permission.WRITE_EXTERNAL_STORAGE",
            "lineNumber" : 7,
            "columnNumber" : 81,
            "fileName" : "/home/suchakra/projects/Our_chat/app/new/AndroidManifest.xml",
            "excerpt" : "<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n    xmlns:tools=\"http://schemas.android.com/tools\"\n    package=\"com.example.ourchat\">\n\n    <uses-permission android:name=\"android.permission.INTERNET\" />\n    <uses-permission android:name=\"android.permission.WRITE_EXTERNAL_STORAGE\" /> /* <===  */ \n    <uses-permission android:name=\"android.permission.RECORD_AUDIO\" />\n    <uses-permission android:name=\"android.permission.FINE_LOCATION\" />\n    <uses-permission android:name=\"android.permission.COARSE_LOCATION\" />\n\n"
          },
          {
            "sample" : "android.permission.WRITE_EXTERNAL_STORAGE",
            "lineNumber" : 7,
            "columnNumber" : 81,
            "fileName" : "/home/suchakra/projects/Our_chat/app/src/main/AndroidManifest.xml",
            "excerpt" : "<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n    xmlns:tools=\"http://schemas.android.com/tools\"\n    package=\"com.example.ourchat\">\n\n    <uses-permission android:name=\"android.permission.INTERNET\" />\n    <uses-permission android:name=\"android.permission.WRITE_EXTERNAL_STORAGE\" /> /* <===  */ \n    <uses-permission android:name=\"android.permission.RECORD_AUDIO\" />\n\n\n    <application\n        android:name=\".MyApplication\""
          }
        ]
      }
    },
```

### CODE REVIEW

File: src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    } catch {
+      case e: Exception => logger.debug("Exception : ", e)
+    }
+    permissions.toList
+  }
+
+  private def getPermissionDetail(node: AndroidXmlPermissionNode): AndroidPermissionDetailModel = {
+    AndroidPermissionDetailModel(
+      sourceId = node.tag.nameExact(Constants.id).value.headOption.getOrElse(""),
+      occurrence = ExporterUtility.convertIndividualPathElement(node).get
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala:47]: as pathElement returned is `Option`, so let's not `get` directly, Better to return `Some(AndroidPermissionDetailModel)` is pathElement exists and similar for the part where this is getting consumed

Comment (tuxology) [reply src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala:47]: Ah yes! missed this. Fixing now

File: src/main/scala/ai/privado/model/exporter/AndroidPermission.scala [hunk 2]
```diff
@@ -9,10 +9,15 @@ case class AndroidPermissionModel(
...
 )
 
-case class AndroidPermissionDetailModel(sourceId: String, occurrence: DataFlowSubCategoryPathExcerptModel)
+case class AndroidPermissionDetailModel(sourceId: String, occurrences: List[DataFlowSubCategoryPathExcerptModel])
 
 object AndroidPermissionsEncoderDecoder {
 
+  implicit val dataFlowSubCategoryPathExcerptModelDecoder: Decoder[DataFlowSubCategoryPathExcerptModel] =
+    deriveDecoder[DataFlowSubCategoryPathExcerptModel]
+  implicit val dataFlowSubCategoryPathExcerptModelEncoder: Encoder[DataFlowSubCategoryPathExcerptModel] =
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/model/exporter/AndroidPermission.scala:18]: These encoder decoder are allready present at `model/exporter/Dataflow.scala`, just import this at the place where encoding/decoding is required

File: src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala [hunk 3]
```diff
@@ -27,24 +28,42 @@ class AndroidPermissionsExporter(cpg: Cpg, ruleCache: RuleCache) {
...
+      }
+      .toList
+    groupedPermissions
   }
 
-  private def getPermissionDetail(node: AndroidXmlPermissionNode): AndroidPermissionDetailModel = {
-    AndroidPermissionDetailModel(
-      sourceId = node.tag.nameExact(Constants.id).value.headOption.getOrElse(""),
-      occurrence = ExporterUtility.convertIndividualPathElement(node).get
+  private def getPermissionDetail(node: AndroidXmlPermissionNode): Some[AndroidPermissionDetailModel] = {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala:58]:
```suggestion
  private def getPermissionDetail(node: AndroidXmlPermissionNode): Option[AndroidPermissionDetailModel] = {
  
  ExporterUtility.convertIndividualPathElement(node) match {
          case Some(pathElement) => Some(AndroidPermissionDetailModel(node.tag.nameExact(Constants.id).value.headOption.getOrElse(""), List(pathElement))
          case None              => None
        }
  }
```

Hope this makes sense

File: src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala [hunk 4]
```diff
@@ -27,24 +28,42 @@ class AndroidPermissionsExporter(cpg: Cpg, ruleCache: RuleCache) {
       cpg.androidXmlPermissionNode
         .where(_.tag.nameExact(Constants.catLevelOne).valueExact(Constants.sources))
         .foreach(node => {
-          permissions.addOne(
-            AndroidPermissionModel(
-              permissionType = node.permissionType,
-              isUsed = true, // we consider permission to be used always
-              permissionDetail = getPermissionDetail(node)
+          val detail = getPermissionDetail(node).get
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/exporter/AndroidPermissionsExporter.scala:31]:
```suggestion
          getPermissionDetail(node) match {
             case Some(detail) => // Now add to permissions
             case None =>
          }
```

**Pull request merged**